src/broadcast.rs

```diff
@@ -0,0 +1,181 @@
+// OF ANY KIND, either express or implied.
+//
+// See the Licences for the specific language governing permissions and limitations relating to
+// use of the MaidSafe Software.
+
+use std::net::{Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6, UdpSocket};
+
+fn serialise_address(our_listening_address: SocketAddr) -> [u8; 27] {
+    let mut our_details = [0u8; 27];
+    match our_listening_address {
+        SocketAddr::V4(ref v4_address) => {
+            // Leave first byte as 0 to indicate IPv4
+            for i in 0..4 {
+                our_details[i + 1] = v4_address.ip().octets()[i];
+            }
+            our_details[5] = (v4_address.port() >> 8) as u8;
+            our_details[6] = v4_address.port() as u8;
+        },
+        SocketAddr::V6(ref v6_address) => {
+            // Set first byte as 1 to indicate IPv6
+            our_details[0] = 1u8;
+            for i in 0..8 {
+                our_details[(2 * i) + 1] = (v6_address.ip().segments()[i] >> 8) as u8;
+                our_details[(2 * i) + 2] = v6_address.ip().segments()[i] as u8;
+            }
+            our_details[17] = (v6_address.port() >> 8) as u8;
+            our_details[18] = v6_address.port() as u8;
+            our_details[19] = (v6_address.flowinfo() >> 24) as u8;
+            our_details[20] = (v6_address.flowinfo() >> 16) as u8;
+            our_details[21] = (v6_address.flowinfo() >> 8) as u8;
+            our_details[22] = v6_address.flowinfo() as u8;
+            our_details[23] = (v6_address.scope_id() >> 24) as u8;
+            our_details[24] = (v6_address.scope_id() >> 16) as u8;
+            our_details[25] = (v6_address.scope_id() >> 8) as u8;
+            our_details[26] = v6_address.scope_id() as u8;
+        },
+    }
+    our_details
+}
+
+fn parse_address(buffer: &[u8]) -> Option<SocketAddr> {
+    match buffer[0] {
+        0 => {
+            let port: u16 = ((buffer[5] as u16) * 256) + (buffer[6] as u16);
+            let peer_socket = SocketAddrV4::new(Ipv4Addr::new(
+                buffer[1], buffer[2], buffer[3], buffer[4]), port);
+            println!("Received IPv4 address {:?}\n", peer_socket);
+            Some(SocketAddr::V4(peer_socket))
+        },
+        1 => {
+            let mut segments = [0u16; 8];
+            for i in 0..8 {
+                segments[i] =
+                    ((buffer[(2 * i) + 1] as u16) << 8) + (buffer[(2 * i) + 2] as u16);
+            }
+            let port: u16 = ((buffer[17] as u16) << 8) + (buffer[18] as u16);
+            let flowinfo: u32 =
+                ((buffer[19] as u32) << 24) + ((buffer[20] as u32) << 16) +
+                ((buffer[21] as u32) << 8) + (buffer[22] as u32);
+            let scope_id: u32 =
+                ((buffer[23] as u32) << 24) + ((buffer[24] as u32) << 16) +
+                ((buffer[25] as u32) << 8) + (buffer[26] as u32);
+            let peer_socket = SocketAddrV6::new(Ipv6Addr::new(
```

Comment:
Ah right does that mean a node may return an address as a message as opposed to from the udp packet itself? Could prove to be better as it will allow further flexibility